src/fix_ea/fix_ea.c

```diff
@@ -0,0 +1,555 @@
+}
+
+static int
+do_ea_stuff_single(const char *path, const char *attr, u_int64_t flags)
+{
+	int fd = 0, setret, ret = 0;
+	struct xattr_list xlist, afp_list, append_list;
+
+	TAILQ_INIT(&xlist);
+	TAILQ_INIT(&afp_list);
+	TAILQ_INIT(&append_list);
+
+	if ((fd = open(path, O_RDONLY)) < 0) {
+		warn("open");
+		ret = EX_OSERR;
+		goto cleanup;
+	}
+
+	if (get_extended_attributes(fd, &xlist) < 0) {
+		ret = EX_DATAERR;
+		goto cleanup;
+	}
+
+	if (flags & F_CHECK_AFP_EA || flags & F_FIX_AFP_EA) {
+		get_afp_list(&xlist, &afp_list);
+		if ((setret = fix_afp_list(fd, path, flags, &afp_list)) < 0) {
+			ret = EX_DATAERR;
+			goto cleanup;
+		}
+		ret = setret;
+	}
+
+	if (flags & F_APPEND_NULL_ALL || flags & F_APPEND_NULL) {
+		get_append_list(&xlist, &append_list, attr);
+		if ((setret = fix_append_list(fd, path, flags, &append_list)) < 0) {
+			ret = EX_DATAERR;
+			goto cleanup;
+		}
+		ret = setret;
+	}
+
+cleanup:
+	unlink_afp_list(&afp_list);
+	unlink_append_list(&append_list);
+	free_extended_attributes(&xlist);
+	close(fd);
+
+	return (ret);
+}
+
+static int
+fts_compare(const FTSENT * const *s1, const FTSENT * const *s2)
+{
+	return (strcoll((*s1)->fts_name, (*s2)->fts_name));
+}
+
+static int
+do_ea_stuff_recursive(char **paths, const char *attr, u_int64_t flags)
+{
+	int rval = 0;
+	FTS *tree;
+	FTSENT *entry;
+
+	if ((tree = fts_open(paths, FTS_LOGICAL | FTS_NOSTAT, fts_compare)) == NULL) {
+		warn("fts_open");
+		return (EX_OSERR);
+	}
+
+	for (rval = 0;(entry = fts_read(tree)) != NULL;) {
+		switch (entry->fts_info) {
+			case FTS_D:
+			case FTS_F:
+				rval |= do_ea_stuff_single(entry->fts_accpath, attr, flags);
+				break;
+
+			case FTS_ERR:
+				warn("%s: %s", entry->fts_path, strerror(entry->fts_errno));
+				break;
+		}
+	}
+
+	fts_close(tree);
+	return (rval);
+}
+
+int
+main(int argc, char **argv)
+{
+	int ch, setret, ret = 0;
+	char *prog, *path, *rp, *attr;
+	u_int64_t flags = F_NONE;
+
+	path = rp = attr = NULL;
+
+	prog = basename(argv[0]);
+	if (argc < 2)
+		usage(prog);
+
+	while ((ch = getopt(argc, argv, "acCdfn:p:rv")) != -1) {
+		switch (ch) {
+			case 'a':
+				flags |= (F_APPEND_NULL_ALL | F_APPEND_NULL);
+				if (attr != NULL) {
+					free(attr);
+					attr = NULL;
+				}
+				break;
+
+			case 'c':
+				flags |= F_CHECK_AFP_EA;
+				flags &= ~F_FIX_AFP_EA;
+				break;
+
+			case 'C':
+				flags |= F_DRY_RUN;
+				break;
+
+			case 'd':
+				flags |= F_DEBUG;
+				break;
+
+			case 'f':
+				flags |= F_FIX_AFP_EA;
+				flags &= ~F_CHECK_AFP_EA;
+				break;
+
+			case 'n':
+				attr = strdup(optarg);
+				flags |= F_APPEND_NULL;
+				flags &= ~F_APPEND_NULL_ALL;
+				break;
+
+			case 'r':
+				flags |= F_RECURSIVE;
+				break;
+
+			case 'v':
+				flags |= F_VERBOSE;
+				break;
+
+			default:
+				usage(prog);
+		}
+	}
+
+	argc -= optind;		
+	argv += optind;
+
+	if (!isatty(STDIN_FILENO)) {
+		ssize_t nread;
+		static char pathbuf[PATH_MAX];
+
+		if ((nread = read(STDIN_FILENO, pathbuf, sizeof(pathbuf))) > 0) {
```

Comment:
This assumes that you get exactly one path line from STDIN? Also, no guarantee it'll be terminated with `\n` or `\0`.
TBH, I'd rather go back to `fgets()` here or `getdelim()` here to avoid complex calculations.

Reply:
This again, is a preference choice and does in fact get 0 terminated. there is no error here.

Reply:
Can you, please, elaborate what usage scenario is expected here? Seems I don't understand, what would be input data.